### PR TITLE
Validate divisor params in lr_scheduler constructors (#177833)

### LINKED ISSUE #193682: LinearLR with total_iters=0 applies start_factor permanently instead of acting as a no-op
### 🐛 Describe the bug

`LinearLR` is documented to interpolate the lr factor from `start_factor` to `end_factor` across `total_iters` steps. With `total_iters=0` the interpolation window is empty, so the schedule is already complete at epoch 0 and the factor should be `end_factor` (default 1.0, i.e. a no-op). Instead, `start_factor` is applied at epoch 0 and never undone — the run trains at `start_factor * base_lr` (default 1/3) forever, silently:

```python
import torch
from torch.optim.lr_scheduler import LinearLR, PolynomialLR

opt = torch.optim.SGD([torch.nn.Parameter(torch.zeros(1))], lr=1.0)
sched = LinearLR(opt, total_iters=0)
for _ in range(3):
    opt.step(); sched.step()
    print(opt.param_groups[0]["lr"])
# 0.3333333333333333
# 0.3333333333333333
# 0.3333333333333333   <- stuck at start_factor * lr for the whole run

# contrast: PolynomialLR handles the same degenerate window correctly
opt2 = torch.optim.SGD([torch.nn.Parameter(torch.zeros(1))], lr=1.0)
sched2 = PolynomialLR(opt2, total_iters=0)
for _ in range(3):
    opt2.step(); sched2.step()
    print(opt2.param_groups[0]["lr"])
# 1.0, 1.0, 1.0
```

### Cause

In `LinearLR.get_lr`:

```python
if self.last_epoch == 0:
    return [group["lr"] * self.start_factor for group in self.optimizer.param_groups]

if self._is_initial or self.last_epoch > self.total_iters:
    return _param_groups_val_list(self.optimizer, "lr")
```

Epoch 0 applies `start_factor` unconditionally; with `total_iters=0` every subsequent epoch satisfies `last_epoch > total_iters` and freezes the lr, so the incremental interpolation that would normally walk the factor up to `end_factor` never executes. `_get_closed_form_lr` divides by `total_iters`, so the closed-form path raises `ZeroDivisionError` for the same config.

### Why this comes up

`total_iters=0` isn't something you type deliberately — it arrives programmatically, e.g. `LinearLR(opt, total_iters=warmup_steps)` where computed `warmup_steps` is 0 for a small run or a resume. The natural reading is "zero-length warmup = no warmup"; the actual behaviour is a permanent 3x lr reduction with no warning, which is easy to misattribute to the model.

### Related

This is the `LinearLR` sibling of #192655 (`ConstantLR` with `total_iters <= 0` scales permanently), with a different mechanism — there the restore branch is unreachable, here the interpolation loop is empty while epoch 0 still applies the start factor. Also adjacent to #189374 (divisor validation) and the semantics RFC #168044. Whichever way the RFC lands — validate-and-reject or treat-as-no-op — `LinearLR` needs one of the two; today it does the third, worst thing (silent permanent scaling), and `PolynomialLR` shows the no-op behaviour is already the norm elsewhere in the module.

Happy to send a PR aligned with whichever direction the maintainers prefer.

### Versions

torch 2.13.0+cpu (also verified against `main` source), Python 3.11.9, Windows.

### PR DESCRIPTION
**Pull request by Functionhx**
## Issues

Fixes #177833
Fixes #193682

Related: #192655

## Summary

Learning-rate schedulers whose formulas divide or take a modulo by constructor parameters accepted non-positive values. This change validates:

- `StepLR.step_size`
- `CosineAnnealingLR.T_max`
- `PolynomialLR.total_iters` and `LinearLR.total_iters`
- `CyclicLR.step_size_up` and `step_size_down`
- `OneCycleLR.div_factor` and `final_div_factor`
- `ConstantLR.factor`

Invalid values now raise a `ValueError` at construction that names the parameter and value. `CyclicLR` validates its step sizes before updating optimizer parameter groups, so failed construction leaves the optimizer reusable.

This follows the approach and review feedback from the stale #178295.

## Root cause and user impact

The parameters above were used as divisors or modulo bases without constructor validation. The previous behavior was not uniform:

- Zero-valued divisors could fail later with `ZeroDivisionError` or invalid arithmetic.
- `LinearLR(total_iters=0)` silently applied `start_factor` permanently during normal chainable stepping; its closed-form path failed with `ZeroDivisionError`.
- `StepLR(step_size=-1)` silently decayed the learning rate on every step because every integer is divisible by `-1`.

The new checks therefore prevent both delayed exceptions and silent incorrect schedules.

## Scope

`ConstantLR.total_iters=0` is tracked separately by #192655. Zero may reasonably mean a no-op there, so this PR does not choose between no-op and rejection semantics for that parameter.

## BC-breaking?

Yes, narrowly: configurations with invalid non-positive values now fail during construction. Some previously failed only when stepped, while `LinearLR(total_iters=0)` and negative `StepLR.step_size` had observable but incorrect behavior. Valid positive configurations are unchanged.

## Test plan

- Added negative tests for zero and negative values with parameter-specific errors.
- Added positive-boundary coverage for minimal valid values.
- Added regression coverage that failed construction does not mutate optimizer parameter groups.
- Targeted scheduler tests: 27 passed.
- `ruff check torch/optim/lr_scheduler.py test/optim/test_lrscheduler.py`

## Checklist

- [x] Added or updated tests
- [x] Checked Python lint
- [ ] Updated documentation (not applicable)
- [ ] Included benchmark results (not applicable)

### CODE REVIEW

**Comment by Functionhx** [comment]
CI appears to be waiting for first-time contributor workflow approval. Could a maintainer approve the pending workflow runs? Thanks.

**Review by shashvat-singham — COMMENTED** [review]
This is a good sweep — validating at construction rather than letting it surface as a `ZeroDivisionError` at `step()` time is clearly the right call, and the coverage of `step_size` / `T_max` / `step_size_up` / `step_size_down` / `div_factor` / `final_div_factor` looks complete to me.

Two notes from probing the same surface.

**1. `ConstantLR.total_iters` is missed.** The PR validates `total_iters` on `LinearLR` and `PolynomialLR`, and validates `factor` on `ConstantLR`, but `ConstantLR.total_iters` gets no check. It has the same parameter and the same failure mode:

```python
opt = torch.optim.SGD([p], lr=1.0)
sched = ConstantLR(opt, total_iters=0)
for _ in range(4):
    opt.step(); sched.step()
    print(opt.param_groups[0]["lr"])
# 0.3333, 0.3333, 0.3333, 0.3333   <- factor applied permanently
```

That's #192655. It doesn't raise `ZeroDivisionError` (the restore branch is simply unreachable rather than dividing), so it may have fallen outside the "divisor params" framing — but it's the same user-visible bug as the `LinearLR` case this PR does fix, in the sibling class. Adding

```python
if total_iters < 0:
    raise ValueError(f"total_iters must be non-negative, but got {total_iters}")
```

to `ConstantLR.__init__` would close it. Worth deciding whether `total_iters=0` should be rejected or treated as a no-op there, since unlike the divisor cases zero is arguably meaningful ("apply the factor for zero steps" = don't apply it).

**2. This PR fixes #193682.** `LinearLR(total_iters=0)` currently applies `start_factor` permanently (1/3 the intended lr for the whole run, silently — it's labeled `module: correctness (silent)`), and the new `total_iters <= 0` check makes it a construction-time error instead. Worth adding to the PR description so the issue closes with it, and so the reviewer sees this isn't only about `ZeroDivisionError` ergonomics — it also removes a silent wrong-result path.

Small thing: the negative case is worth a word in the description too. `StepLR(step_size=-1)` didn't raise before this PR — it silently decayed on *every* step, since `n % -1 == 0` for all `n`. So the `(0, -1)` parametrisation in the new test is catching two genuinely different pre-existing behaviours, not just one.

Verified all of the above against `main` (torch 2.13.0+cpu) before commenting.

**Comment by diputs-sudo** [comment]
@Functionhx 

Thanks for putting this together. I tested the candidate implementation against the existing `torch.optim.lr_scheduler` behavior locally. I agree with the general direction of validating these parameters explicitly rather than allowing invalid configurations to fail later through arithmetic errors, but i found a few things i think we should address before merging this.

1. The BC impact for `total_iters=0` is broader than the PR description currently suggests.

in particular, `LinearLR(total_iters=0)` does not raise during normal chainable stepping today: 

```python
import torch 

p = torch.nn.Parameter(torch.tensor(1.0))
opt = torch.optim.SGD([p], lr=1.0)
sched = torch.optim.lr_scheduler.LinearLR(
    opt, start_factor=0.5, total_iters=0
)

print("initial:", opt.param_groups[0]["lr"])
for i in range(3):
    opt.step()
    sched.step()
    print(f"after step {i + 1}:", opt.param_groups[0]["lr"])

```

Current behavior: 

```
initial: 0.5
after step 1: 0.5
after step 2: 0.5
after step 3: 0.5
```

so this configuration currently has observable behavior which is that the `start_factor` is applied during construction and remains applied under normal chainable stepping.

The closed-from path, exercised with `scheduler.step(0)`, is a seprarate case and raises `ZeroDivisionError`:

```python
p = torch.nn.Parameter(torch.tensor(1.0))
opt = torch.optim.SGD([p], lr=1.0)
sched = torch.optim.lr_scheduler.LinearLR(
    opt, start_factor=0.5, total_iters=0
)
sched.step(0) # ZeroDivisionError
```
(`step(epoch)` is deprecated, but it is useful here specifically to exercise the existing closed-form path.)

I think rejecting `total_iters=0` may still be the right policy, but this means the change is not universally just move an existing `ZeroDivisionError` to construction time. For LinearLR it also replaces an existing silent LR behavior with a constructor error. The BC section should probably make that distinction explicit.

2. `CyclicLR` validation current occurs after optimizer mutation.

A minimal reproducer: 

```python
import torch

p = torch.nn.Parameter(torch.tensor(1.0))
opt = torch.optim.SGD([p], lr=0.1)

try: 
    torch.optim.lr_scheduler.CyclicLR(
        opt,
        base_lr=0.01,
        max_lr=0.1,
        step_size_up=0,
    )
except Exception as e: 
    print(type(e).__name__, e)

print("optimizer lr:", opt.param_groups[0]["lr"])
```

After the failed construction,  the optimizer LR is: 

```
optimizer lr: 0.01
```

rather than the original `0.1`.

I confirmed that this mutation already occurs in the existing implementation, so this is not a regression introduced by this PR. But since this PR is adding explicit constructor validation for `step_size_up` / `step_size_down`, I think those checks should happen before `CyclicLR` mutates the optimizer parameter groups.

The failed test validation direction makes sense to me. I would mainly suggest: 
- expanding the BC description to see cases that currently silently produce LR behavior from cases that only fail later 
- move the `CyclicLR` validation before optimizer mutation 
- adding regression coverage that invalid construction does not mutate optimizer state and that valid scheduler behavior remains the same/

**Comment by Functionhx** [comment]
Thanks for the feedback! Addressed in aca1236. PTAL!
